include/envoy/server/admin.h

```diff
@@ -15,6 +16,34 @@
 namespace Envoy {
 namespace Server {
 
+class AdminStream {
+public:
+  virtual ~AdminStream() {}
+
+  /**
+   * @param end_stream set to false for streaming response. Default is true, which will close
+   * connection on socket
+   */
+  virtual void setEndStreamOnComplete(const bool& end_stream) PURE;
+
+  /**
+   * @param cb callback to be added to the list of callbacks invoked by onDestroy() when connection
+   * is dropped
+   */
+  virtual void addOnDestroyCallback(std::function<void()> cb) PURE;
+
+  /**
+   * @return Http::StreamDecoderFilterCallbacks* to be used by the handler to get socket for data
+   * streaming
+   */
+  virtual const Http::StreamDecoderFilterCallbacks* getDecoderFilterCallbacks() const PURE;
```

Comment:
Can you either make a guarantee that the return-value is non-null by returning a ref, or documenting that nullptr might be an actual return-value?

I think that in the actual usage model, there's no way for null to be returned in the context of an admin callback to which the stream object is passed.  However, I can see that AdminFilter does have these null on construction.

So what's the correct way to use this, and getRequestHeaders()?

One idea is you can make this type-safe by having an StreamImpl object which contains a pointer to the AdminFilter as a delegate, and implements these methods with guarantees implied by returning const ref for the callbacks and requestHeaders, know that at the time is is called, those are guaranteed non-null.

In that case you wouldn't multiply inherit from AdminFilter to AdminStream.

This is a slightly more complex implementation, but a much simpler-to-describe behavior.  With you have, I think you can't make the guarantee without complicated language, and then the caller needs to check for nullptr, creating hard-to-test paths through the code.

WDYT?

Reply:
I do not feel I am familiar enough with the behavior to decide whether we can guarantee a non-null.  Maybe someone else can comment on this. Can you be more explicit about the suggested solution?

Reply:
Another question - there is coverage missing here, but most of it will be covered by the Hystrix test. Can we skip the coverage testing now?

Reply:
RE guarantees: look at the call-site for runCallback in AdminFilter::onComplete, and you'll see that there has to be non-null request-headers there (there's a RELEASE_ASSERT).  At that point you could create the AdminStream object on the stack with the non-null request-headers ref and pass it to runCallback, rather than 'this' (having cut the multiple-inheritance relationship).

It looks to me also that callbacks_ has to be non-null at that point as well, or the call to encodeHeaders will segv.

Reply:
re code-coverage, ideally each level of abstraction should have its own covering unit tests, and not rely on some other complex system to test it.  Is there some reason why this would be hard to test in this case?

Reply:
No, I suggest: 
  1. have AdminStream be a pure base class, not part of a multi-inheritance mesh.
  2. have the callbacks & req-headers be const refs member vars of AdminStreamImpl
  3. construct AdminStreamImpl in AdminFilters::onComplete where it's clear you have those
      available non-null.

Reply:
Re code-coverage: it will be hard to test the onDestroy() code with added callbacks. I could use addHandler() to add a new handler that adds a callback. The two relevant tests are admin_test and integration_admin_test. In admin_test I have an admin object but no requests, so onDestroy() is not called.  In integration_admin_test onDestroy() is called but I don't have an admin object.
I may be missing a simple solution.. it looks to me that it is complicated to build such test, but hystrix_test should cover it, that's why I asked.

Reply:
I understand there are not already existing tests, but I'm still not clear why it's difficult to make one.

I definitely think it's better to cover each level of abstraction on its own and not wait for its application with a higher level integration to be able to test it.